test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+          assert.strictEqual(session, fakeSession);
+          assert.strictEqual(session.type, 'readonly');
+          assert.strictEqual(created, true);
+        });
+      });
 
-    it('should return a trace with the function name', function() {
-      fakeFunction = 'myFunction';
+      it('should fail in creating a read only session', function() {
+        let deleteCalled = false;
+        const fakeSession = {
+          create: () => Promise.reject('Error'),
+          delete: () => {
+            deleteCalled = true;
+            return Promise.resolve();
+          },
+        };
 
-      var expected = `Session leak detected!\n    at ${fakeFunction} (${file})`;
-      var actual = SessionPool.formatTrace_(fakeTrace);
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
 
-      assert.strictEqual(expected, actual);
+        return sessionPool.createReadSession().catch(error => {
+          assert.strictEqual(error, 'Error');
+          assert.strictEqual(deleteCalled, true);
+        });
+      });
     });
-  });
 
-  describe('available', function() {
-    it('should return the number of available sessions', function() {
-      sessionPool.reads_ = [{}];
-      sessionPool.writes_ = [{}, {}];
+    describe('createWriteSession_', function() {
+      it('should create a write session', function() {
+        const fakeSession = {
+          create: () => Promise.resolve(),
+        };
 
-      assert.strictEqual(sessionPool.available(), 3);
-    });
-  });
+        sessionPool.session_ = () => fakeSession;
 
-  describe('borrowed', function() {
-    it('should return the number of borrowed sessions', function() {
-      sessionPool.borrowed_ = [{}, {}];
+        sessionPool.createTransaction_ = function(session) {
+          assert.strictEqual(session, fakeSession);
+          return Promise.resolve();
+        };
 
-      assert.strictEqual(sessionPool.borrowed(), 2);
-    });
-  });
+        return sessionPool.createWriteSession().then(function(session) {
+          assert.strictEqual(session, fakeSession);
+          assert.strictEqual(session.type, 'readwrite');
+        });
+      });
 
-  describe('close', function() {
-    it('should set isOpen to false', function() {
-      sessionPool.isOpen = true;
-      sessionPool.close();
-      assert.strictEqual(sessionPool.isOpen, false);
-    });
+      it('should fail in creating a read/write session', function() {
+        let deleteCalled = false;
+        const fakeSession = {
+          create: () => Promise.reject('Error'),
+          delete: () => {
+            deleteCalled = true;
+            return Promise.resolve();
+          },
+        };
+
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
 
-    it('should emit a close event', function(done) {
-      sessionPool.on('close', done);
-      sessionPool.close();
+        return sessionPool.createWriteSession().catch(error => {
```

Comment:
Need to add a success callback that throws an error to prevent false positives.

Reply:
fixed it